src/NuGet.Core/NuGet.Packaging/Signing/Archive/SignedPackageArchiveIOUtility.cs

```diff
@@ -174,60 +183,72 @@ internal static SignedPackageArchiveMetadata ReadSignedArchiveMetadata(BinaryRea
                 reader.BaseStream.Seek(offset: 8, origin: SeekOrigin.Current);
 
-                var localHeaderOffset = reader.ReadUInt32();
+                centralDirectoryMetadata.OffsetToFileHeader = reader.ReadUInt32();
 
-                var filename = reader.ReadBytes(filenameLength);
-                var filenameString = Encoding.ASCII.GetString(filename);
-                if (string.Equals(filenameString, _signingSpecification.SignaturePath))
+                if (centralDirectoryMetadata.OffsetToFileHeader < metadata.StartOfFileHeaders)
                 {
-                    hasFoundSignatureEntry = true;
+                    metadata.StartOfFileHeaders = centralDirectoryMetadata.OffsetToFileHeader;
+                }
 
-                    metadata.SignatureCentralDirectoryHeaderPosition = possibleSignatureCentralDirectoryRecordPosition;
-                    metadata.SignatureCentralDirectoryHeaderSize = 46 + filenameLength + extraFieldLength + fileCommentLength;
+                var filename = reader.ReadBytes(filenameLength);
+                centralDirectoryMetadata.Filename = Encoding.ASCII.GetString(filename);
+                centralDirectoryMetadata.HeaderSize = 46 + filenameLength + extraFieldLength + fileCommentLength;
 
-                    metadata.SignatureLocalFileHeaderPosition = localHeaderOffset;
+                try
+                {
+                    SeekReaderForwardToMatchByteSignature(reader, BitConverter.GetBytes(CentralDirectoryHeaderSignature));
+                    possibleSignatureCentralDirectoryRecordPosition = reader.BaseStream.Position;
+                }
+                catch
+                {
+                    isReadingCentralDirectoryRecords = false;
+                }
+                centralDirectoryRecords.Add(centralDirectoryMetadata);
+            }
 
-                    // Go to local file header
-                    reader.BaseStream.Seek(offset: metadata.SignatureLocalFileHeaderPosition, origin: SeekOrigin.Begin);
+            var lastCentralDirectoryRecord = centralDirectoryRecords.Last();
+            metadata.EndOfCentralDirectory = lastCentralDirectoryRecord.Position + lastCentralDirectoryRecord.HeaderSize;
 
-                    var fileHeaderSignature = reader.ReadUInt32();
+            // Get missing metadata for central directory records
+            var centralDirectoryRecordsCount = centralDirectoryRecords.Count;
+            for (var i = 0; i < centralDirectoryRecordsCount; i++)
+            {
+                var record = centralDirectoryRecords[i];
 
-                    if (fileHeaderSignature != LocalFileHeaderSignature)
-                    {
-                        throw new InvalidDataException(Strings.ErrorInvalidPackageArchive);
-                    }
+                if (StringComparer.Ordinal.Equals(record.Filename, _signingSpecification.SignaturePath))
+                {
+                    metadata.SignatureIndexInRecords = i;
```

Comment:
should this ensure that the same record wasn't found multiple times? it could just throw if it is.